### DISCUSSION TITLE
Reconstruction plugin - keyboard response rather than button click to finish trial?

### DISCUSSION THREAD

**Discussion by hellenjingyuan** [jspsych/jsPsych#897] [Q&A] [answered]
Thanks a lot for the jspsych-reconstruction plugin! I was wondering if there is a way to use a keyboard response (e.g. pressing the spacebar) to finish the trial, instead of having a Continue button on the screen. Many thanks!

**Comment by jodeleeuw** [comment, marked as answer]
You can do this by modifying the plugin file. It wouldn't be too difficult to do.

You'll need to:

1. Modify the key listener to include the spacebar response:

```js
      valid_responses: [trial.key_increase, trial.key_decrease],
```
becomes 
```js
      valid_responses: [trial.key_increase, trial.key_decrease, 32], 
```
(32 is the key code for the spacebar. You can swap this for a different key or add a parameter to the plugin to control the key.)

2. Update the `after_response` function to do something when the end key is pressed

```js
var after_response = function(info) {

      //console.log('fire');

      var key_i = (typeof trial.key_increase == 'string') ? jsPsych.pluginAPI.convertKeyCharacterToKeyCode(trial.key_increase) : trial.key_increase;
      var key_d = (typeof trial.key_decrease == 'string') ? jsPsych.pluginAPI.convertKeyCharacterToKeyCode(trial.key_decrease) : trial.key_decrease;

      // get new param value
      if (info.key == key_i) {
        param = param + trial.step_size;
      } else if (info.key == key_d) {
        param = param - trial.step_size;
      }
      param = Math.max(Math.min(1, param), 0);

      // refresh the display
      draw(param);
    }
```

becomes

```js
var after_response = function(info) {

      if(info.key == 32){
	   end_trial();
           return; // just to prevent the rest of the code from running.
      }

      //console.log('fire');

      var key_i = (typeof trial.key_increase == 'string') ? jsPsych.pluginAPI.convertKeyCharacterToKeyCode(trial.key_increase) : trial.key_increase;
      var key_d = (typeof trial.key_decrease == 'string') ? jsPsych.pluginAPI.convertKeyCharacterToKeyCode(trial.key_decrease) : trial.key_decrease;

      // get new param value
      if (info.key == key_i) {
        param = param + trial.step_size;
      } else if (info.key == key_d) {
        param = param - trial.step_size;
      }
      param = Math.max(Math.min(1, param), 0);

      // refresh the display
      draw(param);
}
```

3. If you want to remove the button, delete lines 103-105

https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/plugins/jspsych-reconstruction.js#L103-L105

**Reply by hellenjingyuan** [reply]
Thank you so much @jodeleeuw! I implemented the changes you described and ran the script in the browser. When I press the spacebar, the trial doesn't end and I got this error message "Uncaught ReferenceError: end_trial is not defined". The error message refers to line 69 in the modified plugin, which is this part:
```
if(info.key == 32){
        end_trial(); // this is line 69
        return; 
}
```
Any idea? Thanks a lot!

**Reply by jodeleeuw** [reply]
Ah, I think it should be

```js
endTrial()
```

because the function on line 108 is called `endTrial` and not `end_trial()`. My mistake!

https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/plugins/jspsych-reconstruction.js#L108-L131

**Reply by hellenjingyuan** [reply]
That works great. Thank you so much!
